### PR TITLE
Write restart files using the write grid component in UFS

### PR DESCRIPTION
**Pull request by DusanJovic-NOAA**
**Description**

This PR adds new module that enables writing the restart files using the write grid component in UFS. The new module is located in driver/fvGFS/ and is only used by UFS.

Fixes # (issue)

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Please also note
any relevant details for your test configuration (e.g. compiler, OS).  Include
enough information so someone can reproduce your tests.

**Checklist:**

Please check all whether they apply or not
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by bensonr — COMMENTED** [review]
We use the convention of _north_ and _east_ within FMS when referring to staggered variables.  You are already using _east_ and I suggest replacing _south_ with _north_ to keep the conventions the same.

**Comment by DusanJovic-NOAA** [comment]
> We use the convention of _north_ and _east_ within FMS when referring to staggered variables. You are already using _east_ and I suggest replacing _south_ with _north_ to keep the conventions the same.

Done. See https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/244/commits/83f596debedf6ae996144726de5659f53d9439c0

File: driver/fvGFS/fv_ufs_restart_io.F90 [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+   allocate(core_north_var3_names  (1))
+   allocate(core_east_var3_names   (1))
+   allocate(core_center_var3_names (nvar3d_core_center))
+   allocate(core_center_var2_names (1))
+
+   core_north_var3_names(1) = 'u'
+   core_east_var3_names(1) = 'v'
+   n = 1
+   if (.not.atm%flagstruct%hydrostatic) then
+      if (atm%flagstruct%make_nh) then ! hydrostatic restarts dont have these variables
```
Comment (junwang-noaa) [inline driver/fvGFS/fv_ufs_restart_io.F90:122]: I am not sure if I understand correctly. But the lines 123-127 look the same as line 129-133, do we need this if statement?

File: driver/fvGFS/fv_ufs_restart_io.F90 [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+
+   ! ---- fv_core.res
+
+   core_north_var3(:,:,:,1) = Atm%u(isc:iec,jsc:(jsc+ny_c-1),:)
+   core_east_var3 (:,:,:,1) = Atm%v(isc:(isc+nx_c-1),jsc:jec,:)
+
+   n = 1
+   if (.not.atm%flagstruct%hydrostatic) then
+      if (atm%flagstruct%make_nh) then ! hydrostatic restarts dont have these variables
+           ! core_center_var3(n) = 'w'; n=n+1
```
Comment (junwang-noaa) [inline driver/fvGFS/fv_ufs_restart_io.F90:210]: Not sure if I understand this if statement (and comments) . Isn't line 208 said that this is non-hydrostatic? Can we have:
```
if (.not. atm%flagstruct%make_nh) then
 lines 216-220
endif
```

Comment (DusanJovic-NOAA) [reply driver/fvGFS/fv_ufs_restart_io.F90:210]: I copied the logic from:

https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/dev/emc/tools/fv_io.F90#L301-L315

**Review by junwang-noaa — APPROVED** [review]

**Comment by junwang-noaa** [comment]
> > We use the convention of _north_ and _east_ within FMS when referring to staggered variables. You are already using _east_ and I suggest replacing _south_ with _north_ to keep the conventions the same.
> 
> Done. See [83f596d](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/commit/83f596debedf6ae996144726de5659f53d9439c0)

@bensonr would you please review again? Thanks

**Review by bensonr — APPROVED** [review]

**Comment by jkbk2004** [comment]
all tests are done on https://github.com/ufs-community/ufs-weather-model/pull/1633. @bensonr @laurenchilutti can you merge the pr?

**Pull request merged**